xformers/components/attention/core.py

```diff
@@ -210,19 +215,94 @@ def scaled_query_key_softmax(
     return att
 
 
+# 128 is default maxsize
+@lru_cache(maxsize=128)
+def _retrieve_blocksparse(
+    num_heads: int, seq_len: int, block_size: int
+) -> BlockSparseAttention:
+    # Checks if blocksparse object exists in cache
+
+    blocks = seq_len // block_size
+    print("Made uncached blocksparse")
+    layout_fill = torch.ones((num_heads, blocks, blocks), dtype=torch.long)
+    return BlockSparseAttention(layout=layout_fill, block_size=block_size, causal=True)
+
+
+def blocksparse_attention(
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    dropout: Optional[torch.nn.Module] = None,
+    block_size: int = 128,
+) -> torch.Tensor:
+
+    orig_dim = q.dim()
+    seq_len = q.shape[-2]
+    # Layout head dimension: 1 or batch size (q.shape[0])
+    layout_heads = 1
+
+    # TODO perhaps add functionality to pad qkv if sequence length is not divisible by block size?
+    assert seq_len % block_size == 0, "Sequence length must be divisible by block size"
```

Comment:
I think in the future we should just have a custom set of operators (`sddmm` / `softmax` / spmm`) which work on causal structure. This would enable for fast execution without being dependent on block sizes, and should be faster than blocksparse as we wouldn't need to index on a set of (unused) indices